_posts/2018-03-01-Rust-1.24.1.md

```diff
@@ -0,0 +1,257 @@
+
+If you have a previous version of Rust installed via rustup, getting Rust
+1.24.1 is as easy as:
+
+```bash
+$ rustup update stable
+```
+
+If you don't have it already, you can [get `rustup`][install] from the
+appropriate page on our website, and check out the [detailed release notes for
+1.24.1][notes] on GitHub.
+
+[install]: https://www.rust-lang.org/install.html
+[notes]: https://github.com/rust-lang/rust/blob/stable/RELEASES.md#version-1241-2018-03-01
+
+## What's in 1.24.1 stable
+
+As you know, we tend not to release point releases. In this case, we had a few regressions in 1.24.0.
+Each of them individually was not a "hair on fire" moment, but there's two reasons we decided to
+produce a point release. First, collectively, they were big enough to consider as a whole. Second, we
+want our attitude towards regressions to be "if we're not sure, default to reverting." Given that
+it was not clear that we *shouldn't* release a point release, we decided to err on the side of
+caution.
+
+A quick summary of the changes: there are four.
+
+* Do not abort when unwinding through FFI
+* Emit UTF-16 files for linker arguments on Windows
+* Make the error index generator work again
+* Cargo will warn on Windows 7 if an update is needed.
+
+In general, if you haven't run into any build issues, the "do not abort when unwinding through FFI"
+change is the only one you should care about. The others are also important, but if 1.24.0 has been
+working well for you, this is the only change that affects you. We plan on bringing this behavior back
+in 1.25 or 1.26, depending on how smoothly the new strategy goes.
+
+With that, let's dig into the details!
+
+### Do not abort when unwinding through FFI
+
+Quoting [the 1.24 annoucement](https://blog.rust-lang.org/2018/02/15/Rust-1.24.html):
+
+> There’s one other change we’d like to talk about here: undefined behavior.
+> Rust generally strives to minimize undefined behavior, having none of it in
+> safe code, and as little as possible in unsafe code. One area where you could
+> invoke UB is when a panic! goes across an FFI boundary. In other words, this:
+
+```rust
+extern "C" fn panic_in_ffi() {
+    panic!("Test");
+}
+```
+
+> This cannot work, as the exact mechanism of how panics work would have to
+> be reconciled with how the "C" ABI works, in this example, or any other ABI
+> in other examples.
+>
+> In Rust 1.24, this code will now abort instead of producing undefined behavior.
+
+We have reverted this behavior, and 1.24.1 will reintroduce the undefined behavior.
+Fundamentally, this is a soundness fix, and we still plan on reintroducing the
+abort, but are going to be rolling it out a bit more slowly, and with a different
+implementation.
+
+It started with [a bug filed against the `rlua`
+crate](https://github.com/chucklefish/rlua/issues/71). `rlua` is a package
+that provides high level bindings between Rust and the [Lua programming
+language](https://www.lua.org/).
```

Comment:
It is probably worth explicitly acknowledging here that the `rlua` implementation depends on undefined behavior that just *happens* to work on all known OSes as of Rust 1.23. (That is correct, right?)

Reply:
> It is probably worth explicitly acknowledging here that the rlua implementation depends on undefined behavior that just happens to work on all known OSes as of Rust 1.23. (That is correct, right?)

I'm going to attempt to explain the situation so that if I get something wrong somebody knowledgeable will correct me, because I really want to know the precise answer as well!

I think the question of whether what I was doing was / is UB is entirely a question of Rust specification.  As far as I am aware, there wasn't much documentation or decisions on whether or not a longjmp as implemented on unixes was something that should work across Rust frames, but it was clear that if the Rust frame being jumped over was "trivial", this "should not" cause any issues.  The problem only arose from a combination of two factors, which is that longjmp on windows was implemented using SEH and I was not aware of this, and Rust's addition of automatic exception handling on extern fn boundaries.

Now, I think it could be argued that since there was definite documentation on unwinding through rust being disallowed, and SEH is certainly the sort of unwinding that was being referred to, rlua with rust 1.23.0 on windows was relying on UB that just happened to work.  You could possibly also argue about the precise definition of "unwinding" and say that rlua was relying on UB categorically, I'm honestly not sure whether unwinding refers only to things like exception handling or also to simple stack manipulation.  I think there was consensus in the bug report, though, that disallowing interacting with C APIs that longjmp is not great for Rust's C interop story, so it SEEMED that the decision was made that longjmp is something that SHOULD be supported, at least in the precise limited way that rlua was trying to use it, because otherwise there are quite a few C APIs that just become extremely painful to use, and would REQUIRE extremely complex C wrapper APIs.  There was also talk about supporting interop with exceptions and unwinding more generally going forward, and doing a better job of checking soundness conditions on APIs like Lua's, but obviously that wasn't going to be part of the current fix.

So, the end result is that.. you could argue that rlua was relying on UB or at least unspecified behavior but is now NOT relying on UB, simply because it was decided that it shouldn't be UB.  It's kind of a messy question!

If I've gotten any of that wrong let me know, because I don't want to misrepresent the intentions of the people who ACTUALLY make these decisions.

Reply:
I know I'm being verrrry hand wavey with the previous explanation, and if you feel skeptical about my reasoning, I don't blame you.

We were ultimately aware of how weird and potentially bad longjmping over Rust was, and `rlua` had an old issue with a pretty long discussion about potential fixes in it.  The problem though, ultimately, was that the solutions were just really bad, and not being able to do this makes Rust / C interop feel very limited.

I hesitate even still to call the rust behavior a "bug" per se, it felt to me like the bug report was more a process to discuss an edge case of C interop, and decide what Rust should do going forward.

Reply:
@kyren 

I would assert that UB is *always* a matter of language specification! (C++ folks will get touchy on this point; anything that is specified by the OS is merely *unspecified*, not *undefined*.)

Since Rust didn't provide a *guarantee* of a particular behavior, I think I would indeed consider the situation (as I understand it) UB. But I also wouldn't necessarily consider that a "bug", because UB just means that *the Rust language* doesn't impose a behavioral requirement. In this case, it seems that cross-platform behavior was consistent and correct, even without a formal specification of the desired or actual behavior.

I would also say that, regardless of whether we call this "UB" or not, the ideal path would be to make Rust *guarantee* that this use-case will work in the future, because it seems clear that it's a valuable behavior. And, yes, that would be equivalent to saying that in 1.24.0 and earlier, it was UB, *even though* it works correctly in 1.23 and earlier; and that in 1.24.1+ it won't be, "simply because it was decided that it shouldn't be UB."

Reply:
Oh, okay we're on the same page then.  I've had this conversation a few times, and sometimes when people say UB they mean it as a synonym for "definitely should not do it" or "the specification explicitly states that it is UB" or simply "you have a bug" and not "the specification does not explicitly allow it or does not address it yet".  I was attempting to draw a distinction between "unspecified" and "undefined" that lumped in decisions yet to be made into "unspecified", which I guess is not really accurate.  I think in C++ parlance they sometimes use the words unspecified or "platform specified" to delineate between behavior that they have actively decided compilers are allowed to do whatever they want, and behavior that is simply not part of the specification.

Communicating about this is hard, man!

Reply:
@kyren Yeah, I think we are on the same page. I didn't want to get too much into my personal gripes with the state of affairs in C and C++, but suffice to say that I think the attitude that "if you have UB you deserve whatever bad thing happens to you" is unhelpful at best. I have also come to believe that it's somewhat disingenuous and ahistorical; supercat's comments [here](https://stackoverflow.com/q/32132574/1858225) were particularly enlightening to me. (They commented on almost every answer; their comments are, oddly, more valuable to me than their answer.)

You're right that "undefined" and "unspecified" are different, especially in C++ world. But the C++ standard explicitly states that "compilers are allowed to do whatever they want" and "behavior that is simply not part of the specification" is generally *the same thing* unless explicitly stated otherwise. I think that's...not a great approach, especially when the committee has little to no interest or incentive to *reduce* the amount of undefined behavior in the language over time.

In Rust things really are a little fuzzier, though, since we don't have a formal language specification the way C++ does. Then again, C++ got by without one for over a decade, which the UB zealots (sorry, can't help myself) conveniently tend to forget...